### PR TITLE
Step5 - deeplink

### PR DESCRIPTION
**Pull request by sagilorber**
DeepLink added. it caches link with the pattern: 
https://www.themoviedb.org/movie/.* when .* has the movie id.
otherwise  it will launch the airing now list.

deeplink

### CODE REVIEW

File: app/src/main/java/com/example/slorber/moviefiend/DeepLinkHandlerActivity.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        Intent backIntent = new Intent(this, MainActivity.class);
+        backIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
+
+        Intent intent = new Intent(this, MovieDetailActivity.class);
+        intent.putExtra(DEEPLINKEXTRA, getIntent().getData().toString());
+        final PendingIntent pendingIntent = PendingIntent.getActivities(this, 1, new Intent[] { backIntent, intent }, PendingIntent.FLAG_ONE_SHOT);
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/DeepLinkHandlerActivity.java]: Any reason not to use `TaskStackBuilder`?

File: app/src/main/java/com/example/slorber/moviefiend/MainActivity.java [hunk 2]
```diff
@@ -75,13 +59,35 @@ public void onLoadFinished(Loader<List<Movie>> loader, List<Movie> data) {
...
+//                    Intent intent2 = new Intent(this, MovieDetailActivity.class);
+//                    Bundle b = new Bundle();
+//                    b.putParcelable(MovieDetailActivity.EXTRA_MOVIE, data.get(i));
+//                    intent2.putExtras(b);
+//                    startActivity(intent2);
+//                    break;
+//                }
+//            }
+//        }
+//    }
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MainActivity.java]: please remove all commented-out code

File: app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java [hunk 3]
```diff
@@ -10,52 +11,132 @@
...
+    public static final String EXTRA_ID = "id";
+    private static final String URL = "http://api.themoviedb.org";
     public static final String EXTRA_MOVIE = "Movie";
     private static final String IMAGE_URL = "http://image.tmdb.org/t/p/w500/";
-    private Movie movie;
+    private Movie mMovie;
+    private RatingView ratingView;
+    private MovieDetailsView movieDetailsView;
+    private ImageView imageView;
+    private boolean launchedFromDeepLink = false;
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java]: `m` prefix for member variables.

File: app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java [hunk 4]
```diff
@@ -10,52 +11,132 @@
...
-        Bundle b = this.getIntent().getExtras();
-        if (b != null) {
-            movie = b.getParcelable(EXTRA_MOVIE);
+        if (getIntent().getParcelableExtra(EXTRA_MOVIE) != null) {
+            setMovie((Movie) getIntent().getExtras().getParcelable(EXTRA_MOVIE));
+
+        } else {
+            checkDeepLink(getIntent());
+            launchedFromDeepLink = true;
         }
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java]: The `MovieDetailActivity` should be initialized in a single way (by passing the `EXTRA_MOVIE`).
This way you won't need to separately handle the deep link scenario.

File: app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java [hunk 5]
```diff
@@ -10,52 +11,132 @@
...
         startActivity(intent);
     }
 
+    @Override
+    public Loader<Movie> onCreateLoader(int id, Bundle args) {
+        return new GetMovieDetailLoader(this, Uri.parse(URL)
+                .buildUpon()
+                .appendPath("3")
+                .appendPath("movie")
+                .appendPath(String.valueOf(args.get(EXTRA_ID)))
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java]: `args.getString` - then you won't need the `String.valueOf`

File: app/src/main/java/com/example/slorber/moviefiend/GetMovieDetailLoader.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+ * TODO: Write Javadoc for GetMovieDetailLoader.
+ *
+ * @author slorber
+ */
+public class GetMovieDetailLoader extends Loader<Movie> implements TMDBApi.SingleMovieListener {
+
+    private Movie mMovie;
+    private String mUrl;
+
+    public GetMovieDetailLoader(Context context, String url) {
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/GetMovieDetailLoader.java]: Makes more sense for it to get a `Uri`

File: app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java [hunk 7]
```diff
@@ -10,52 +11,132 @@
...
+            if (extraUri.getLastPathSegment().contains("-") && TextUtils.isDigitsOnly(extraUri.getLastPathSegment().split("-")[0])) {
+                String[] parts = extraUri.getLastPathSegment().split("-");
+
+                int deepLinkMovieId = Integer.parseInt(parts[0]);
+                Bundle b = new Bundle();
+                b.putInt(EXTRA_ID, deepLinkMovieId);
+                getSupportLoaderManager().initLoader(0, b, this);
+            }
+            else
+                finish();
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java]: We wrap code blocks in curly braces even if there is only one line of code.

File: app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java [hunk 8]
```diff
@@ -10,52 +11,132 @@
...
+//            Intent intent = new Intent(this, MainActivity.class);
+//            intent.addFlags(Intent.FLAG_ACTIVITY_REORDER_TO_FRONT);
+//            startActivity(intent);
+//            finish();
+//        }
+//        else
+//        {
+//            super.onBackPressed();
+//        }
+//    }
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java]: commented out code

File: app/src/main/java/com/example/slorber/moviefiend/TMDBApi.java [hunk 9]
```diff
@@ -53,8 +53,34 @@ public void onErrorResponse(VolleyError error) {
         queue.add(getRequest);
 
     }
+    public void getMovieDetails(Context context, String url, final SingleMovieListener listener) {
+        RequestQueue queue = Volley.newRequestQueue(context);
+        JsonObjectRequest getRequest = new JsonObjectRequest(Request.Method.GET, url, null,
+                new Response.Listener<JSONObject>() {
+                    @Override
+                    public void onResponse(JSONObject response) {
+                        Gson g = new Gson();
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/TMDBApi.java]: No need to instantiate `Gson` for every request.
You can store it as a member variable that will serve the whole class.

File: app/src/main/java/com/example/slorber/moviefiend/TMDBApi.java [hunk 10]
```diff
@@ -53,8 +53,34 @@ public void onErrorResponse(VolleyError error) {
...
+                        Movie movie = g.fromJson(response.toString(), Movie.class);
+                        listener.onMovieFetched(movie);
+                    }
+                },
+                new Response.ErrorListener() {
+                    @Override
+                    public void onErrorResponse(VolleyError error) {
+
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/TMDBApi.java]: format

File: app/src/main/res/layout/activity_deep_link_handler.xml [hunk 11]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (ifeins) [inline app/src/main/res/layout/activity_deep_link_handler.xml]: Is this file needed?

File: app/src/main/java/com/example/slorber/moviefiend/DeepLinkHandlerActivity.java [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        Intent backIntent = new Intent(this, MainActivity.class);
+        Intent intent = new Intent(this, MovieDetailActivity.class);
+        TaskStackBuilder stackBuilder = TaskStackBuilder.create(this);
+        stackBuilder.addParentStack(MainActivity.class);
+        stackBuilder.addNextIntent(backIntent);
+        stackBuilder.addNextIntent(intent);
+        intent.putExtra(DEEPLINKEXTRA, getIntent().getData().toString());
+        final PendingIntent pendingIntent =stackBuilder.getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT);// PendingIntent.getActivities(this, 1, new Intent[] { backIntent, intent }, PendingIntent.FLAG_ONE_SHOT);
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/DeepLinkHandlerActivity.java]: no commented out code

File: app/src/main/java/com/example/slorber/moviefiend/DeepLinkHandlerActivity.java [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+        Intent intent2 = new Intent();
+        try {
+            pendingIntent.send(this, 0, intent2);
+            finish();
+
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/DeepLinkHandlerActivity.java]: format

File: app/src/main/java/com/example/slorber/moviefiend/MainActivity.java [hunk 14]
```diff
@@ -75,13 +59,16 @@ public void onLoadFinished(Loader<List<Movie>> loader, List<Movie> data) {
...
 
     @Override
     public void onLoaderReset(Loader<List<Movie>> loader) {
 
     }
-
+    @Override
+    protected void onNewIntent(Intent intent) {
+        super.onNewIntent(intent);
+    }
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MainActivity.java]: You don't seem to override anything in this method.
Is it needed?

File: app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java [hunk 15]
```diff
@@ -10,52 +11,116 @@
...
         setSupportActionBar((Toolbar) findViewById(R.id.toolbar));
         getSupportActionBar().setDisplayHomeAsUpEnabled(true);
-        Bundle b = this.getIntent().getExtras();
-        if (b != null) {
-            movie = b.getParcelable(EXTRA_MOVIE);
+        getSupportActionBar().setDisplayShowTitleEnabled(false);
+        if (getIntent().getParcelableExtra(EXTRA_MOVIE) != null) {
+            setMovie((Movie) getIntent().getExtras().getParcelable(EXTRA_MOVIE));
+        } else {
+            checkDeepLink(getIntent());
```
Comment (ifeins) [inline app/src/main/java/com/example/slorber/moviefiend/MovieDetailActivity.java]: You can move the loading of the movie to the deep link activity and then this activity can still be initialized with the `EXTRA_MOVIE` variable.

This way if an error occurs during the loading of the movie then it's easier to show an error page or maybe to show the main list page.

File: app/src/main/AndroidManifest.xml [hunk 16]
```diff
@@ -1,41 +1,57 @@
...
+                  android:theme="@android:style/Theme.NoDisplay">
+            <intent-filter>
+                <action android:name="android.intent.action.VIEW"/>
+
+                <category android:name="android.intent.category.DEFAULT"/>
+                <category android:name="android.intent.category.BROWSABLE"/>
+
+                <data
+                    android:host="www.themoviedb.org"
+                    android:pathPattern="/movie/..*"
```
Comment (sagilorber) [inline app/src/main/AndroidManifest.xml]: no need for two dots... i wanted to catch all /movie/something paths. can be done by a single dot.
